### ISSUE TITLE
Query.lazy_loaded_from not populating properly for parent->children relationships

### ISSUE THREAD

**Issue by aklauza** [sqlalchemy/sqlalchemy#5109]
Hi Mike & SQL Alchemy community!

We are trying to use the `before_compile` event in order to add  some hints to the query based on the parent instance and propagate down to lazy loaded queries, but are finding that this new feature added to populate this attribute does not always work as expected. Please see example below to understand:

```python
from typing import List

from production.config.App import App
from sqlalchemy import Column, Integer, String, ForeignKey, event
from sqlalchemy.orm import relationship, Query

from production.models import db



class Son(db.Model):
    __tablename__ = 'sons'

    id: int = Column(Integer, primary_key=True, autoincrement=True)
    name: str = Column(String(50), nullable=False, index=True)
    father_id: int = Column(Integer, ForeignKey('fathers.id'), nullable=False)

    father: 'Father' = relationship('Father')


class Father(db.Model):
    __tablename__ = 'fathers'

    id: int = Column(Integer, primary_key=True, autoincrement=True)
    name: str = Column(String(50), nullable=False, index=True)
    children: List[Son] = relationship('Son')



@event.listens_for(Query, 'before_compile')
def receive_before_compile(query):
    print(f'lazy_loaded_from is: {query.lazy_loaded_from}')
    return query



App().setup_standalone()  # This just does all the sqlAlchemy setup

# This works, the first query issued is not lazy so it's None, the second line is lazy so
# the field lazy_loaded_from contains properly the Son Instance (initiator)
print('Get the father from a Son instance')
one_kid = Son.query.filter_by(id=1).one()
print(one_kid.father)

# If we do it the other way I would expect it to work as well, but the result is none in 
# both cases.
print('\nGet children from Father instance')
a_father = Father.query.filter_by(id=5).one()
print(a_father.children)
```
This a pretty simple relationship case, a one to many relationship where a Father has many children but each Son has only one father.

**When running this I would expect in **BOTH** cases the second `lazy_loaded_from:` print to show the related entity instance. However this only happens in the first case, not the second. See output** below:
```bash
Get the father from a Son instance
lazy_loaded_from is: None
lazy_loaded_from is: <sqlalchemy.orm.state.InstanceState object at 0x0000026E4962DB70>
<Father 1>

Get children from Father instance
lazy_loaded_from is: None
lazy_loaded_from is: None
[<Son 22>, <Son 23>, <Son 24>, <Son 25>, <Son 26>, <Son 27>, <Son 28>, <Son 29>, <Son 30>, <Son 31>]
```
This looks like a bug when setting that attribute. If this is expected, please share an alternative way we can get from within a lazy loaded `before_compile` event the invoquer instance.

Note: We are using Flask-Sqlalchemy wrapper in this example, but that does not really affect any of the details here explained as the `lazy_loaded_from` field and `before_compile` events are managed by SqlAlchemy.

Thank you very much

**Comment by zzzeek** [comment]
hi there -

I haven't read the whole email since I see "before_compile()", "lazy loading" and i do *not* see the "bake_ok" flag, documented [here](https://docs.sqlalchemy.org/en/13/orm/events.html?highlight=before_compile#sqlalchemy.orm.events.QueryEvents.before_compile) this is required so that query caching is disabled for lazy loader queries that hit upon this event hook.

**Comment by zzzeek** [comment]
Correction, bake_ok defaults to False but you need to ensure you are returning a new Query object for the "don't cache" hook to be triggered.  Also you can put bake_queries=False on the relationship():  https://docs.sqlalchemy.org/en/13/orm/relationship_api.html?highlight=bake_queries#sqlalchemy.orm.relationship.params.bake_queries

**Comment by aklauza** [comment]
Thanks for your quick response @zzzeek. One of the solutions do work, but not the other:
- If I set `bake_queries=False` at relationship level it works as I expect it and the `lazy_loaded_from` attr. in the query gets set properly.
- But if I try to do it at the `before_compile` hook it does not work. You mention I need to return a new Query object for it to work, and I think I am doing it in the snippet below but still does not work.
```python
@event.listens_for(Query, 'before_compile', retval=True, bake_ok=False)
def receive_before_compile(query:Query):
    print(f'lazy_loaded_from is: {query.lazy_loaded_from}')
    return query.order_by()  # This should do the trick and create a new object (I checked hashes and they are different)
```
What do you think? Thanks

Note: I am using SqlAlchemy=1.3.12

**Comment by zzzeek** [comment]
OK I've gone back and read specifically what you're looking for and here is what's going on.

1. the lazy_loaded_from attribute was [added](https://docs.sqlalchemy.org/en/13/changelog/changelog_13.html#change-ebf60a359d3a5e8d3c048f04ce5781c9) first and foremost to suit an extremely specific use case in the horizontal sharding extension, which is that for a lazy load operation, a horizontal sharder will want to see what the object we're loading from is, so that it can more intelligently pick the correct shard identifier.   these hooks occur either in the "execute" phase, after the Query has gone through the compilation stage already, or in the "identity lookup" stage which happens outside of the query compile/execute process.

2. the docstring that is underneath lazy_loaded_from, which includes, "custom mapper options", that part is flat out wrong and always has been, because the lazy loader applies the "lazy_loaded_from" after it would have applied any mapper options.

3. the part of the docstring that says "event handlers", is more or less wrong right now but it doesn't even specify what "event handlers" we are talking about.  when the feature was added in 1.2.x, the before_compile() event would have been able to get at this value consistently.   in 1.3, since lazy loading caches the queries, the before_compile() event cannot have access to this attribute because before_compile() is not called for the vast majority of lazy load operations.  additionally, the "bake_ok" thing,which does prevent the query from being "baked" and has the before_compile() event called each time, is nonetheless implemented in such a way that the flow is still similar to a query that would be cached, and the "lazy loaded from" attribute is a post-compile hook so again is not set.   you will note that if your script tries to do the same operation a few more times, the attribute is None for both the many-to-one and one-to-many loads.    so at the moment there are no event handlers that can be used to intercept "lazy_loaded_from" consistently assuming caching is being used, since we lack a "before_query_execute()" event at the moment.     we'll update those incorrect docs here.

4. the next major version of SQLAlchemy is likely to feature a reorganization of statement execution such that for this use case you'd likely use the Core level before_execute() event in order to intercept a Query before it is executed.    That's not available right now.


TL;DR;

1. docs are wrong, fixing

2. turn of bake_queries for now

3. in 1.4, hopefully, you'll use before_execute() event, or something like it

**Comment by aklauza** [comment]
Thanks Mike, that makes sense. We are fine using the `bake_queries=False` for now.

Also, FYI to provide a bit more context of what we are trying to do here and why it's similar to the sharding case:
- We want to implement **Temporal tables querying within sqlAlchemy engine** (it's a relatively new feature for point-in-time querying in SQL server: https://docs.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver15)
- So we could query an object and all its relationships at a point in time:
```python
Father.query_as_of(date(2015,10,25)).filter(Father.name=='Arek').one()
``` 
- To use it we need to add a hint to the queries before they are emmited right after the from statement with an `as_of_date`:
```sql
SELECT * FROM fathers FOR SYSTEM_TIME AS OF '{as_of_date}'
```
- We intercept the first query call above, attach/save the `as_of_date` the user inputs to an object attribute, and then
- whenever a lazy loaded query is issued to related objects, we want to intercept them in `before_compile` to get the `as_of_date` from the original object (using lazy_loaded_from) and add is as hint to the lazy query as well (because we want the relationships from the past as well).

Thanks for your great work @zzzeek!
